# Step 01: Data Understanding & Initial Profiling

## Overview
This notebook performs exploratory profiling across the 5 core CSV datasets for the **Enterprise HR AI** platform:
1. `employee_attrition.csv`
2. `hr_performance_engagement.csv`
3. `occupation_data.csv`
4. `essential_skills.csv`
5. `software_skills.csv`

For each dataset, we assess shape, data types, missing value percentages, duplicate counts, primary key candidates, target class balance (where applicable), and join relationships.


In [1]:
import pandas as pd
import numpy as np
import os

RAW_DIR = os.path.join("..", "data", "raw")
print("Raw Data Directory:", os.path.abspath(RAW_DIR))


Raw Data Directory: C:\Users\komal\Desktop\enterprise_hr_ai\data\raw


---
## 1. Employee Attrition Dataset (`employee_attrition.csv`)
Target dataset containing employee demographic, job, and attrition status.


In [2]:
attr_df = pd.read_csv(os.path.join(RAW_DIR, "employee_attrition.csv"))
print(f"Shape: {attr_df.shape} (Rows: {attr_df.shape[0]}, Cols: {attr_df.shape[1]})")
print("\nColumns & Dtypes:")
print(attr_df.dtypes)

print("\nMissing Values:")
missing = attr_df.isnull().sum()
print(missing[missing > 0] if missing.sum() > 0 else "No missing values.")

print(f"\nDuplicate rows: {attr_df.duplicated().sum()}")
print(f"Unique EmployeeNumbers: {attr_df['EmployeeNumber'].nunique()} / {len(attr_df)}")

print("\n--- Target Class Balance ('Attrition') ---")
target_counts = attr_df['Attrition'].value_counts(dropna=False)
target_props = attr_df['Attrition'].value_counts(normalize=True, dropna=False) * 100
target_summary = pd.DataFrame({'Count': target_counts, 'Percentage (%)': target_props})
print(target_summary)


Shape: (1470, 35) (Rows: 1470, Cols: 35)

Columns & Dtypes:
Age                          int64
Attrition                   object
BusinessTravel              object
DailyRate                    int64
Department                  object
DistanceFromHome             int64
Education                    int64
EducationField              object
EmployeeCount                int64
EmployeeNumber               int64
EnvironmentSatisfaction      int64
Gender                      object
HourlyRate                   int64
JobInvolvement               int64
JobLevel                     int64
JobRole                     object
JobSatisfaction              int64
MaritalStatus               object
MonthlyIncome                int64
MonthlyRate                  int64
NumCompaniesWorked           int64
Over18                      object
OverTime                    object
PercentSalaryHike            int64
PerformanceRating            int64
RelationshipSatisfaction     int64
StandardHours                i

---
## 2. HR Performance & Engagement Dataset (`hr_performance_engagement.csv`)
Contains employee performance ratings, survey engagement, satisfaction, and work-life balance scores.


In [3]:
perf_df = pd.read_csv(os.path.join(RAW_DIR, "hr_performance_engagement.csv"))
print(f"Shape: {perf_df.shape} (Rows: {perf_df.shape[0]}, Cols: {perf_df.shape[1]})")
print("\nColumns & Dtypes:")
print(perf_df.dtypes)

print("\nMissing Values:")
missing = perf_df.isnull().sum()
print(missing[missing > 0] if missing.sum() > 0 else "No missing values.")

print(f"\nDuplicate rows: {perf_df.duplicated().sum()}")
print(f"Unique Employee IDs: {perf_df['Employee ID'].nunique()} / {len(perf_df)}")

print("\nEngagement & Satisfaction Score Distributions:")
print(perf_df[['Engagement Score', 'Satisfaction Score', 'Work-Life Balance Score', 'Current Employee Rating']].describe())


Shape: (2845, 28) (Rows: 2845, Cols: 28)

Columns & Dtypes:
Employee ID                     int64
StartDate                      object
Title                          object
BusinessUnit                   object
EmployeeStatus                 object
EmployeeType                   object
PayZone                        object
EmployeeClassificationType     object
DepartmentType                 object
Division                       object
DOB                            object
State                          object
GenderCode                     object
RaceDesc                       object
MaritalDesc                    object
Performance Score              object
Current Employee Rating         int64
Survey Date                    object
Engagement Score                int64
Satisfaction Score              int64
Work-Life Balance Score         int64
Training Date                  object
Training Program Name          object
Training Type                  object
Training Outcome            

---
## 3. Occupation Data Master (`occupation_data.csv`)
O*NET master reference table mapping SOC job codes to standardized titles and descriptions.


In [4]:
occ_df = pd.read_csv(os.path.join(RAW_DIR, "occupation_data.csv"))
print(f"Shape: {occ_df.shape} (Rows: {occ_df.shape[0]}, Cols: {occ_df.shape[1]})")
print("\nColumns & Dtypes:")
print(occ_df.dtypes)

print("\nMissing Values:")
missing = occ_df.isnull().sum()
print(missing[missing > 0] if missing.sum() > 0 else "No missing values.")

print(f"\nDuplicate rows: {occ_df.duplicated().sum()}")
print(f"Unique O*NET-SOC Codes: {occ_df['O*NET-SOC Code'].nunique()} / {len(occ_df)}")


Shape: (1016, 3) (Rows: 1016, Cols: 3)

Columns & Dtypes:
O*NET-SOC Code    object
Title             object
Description       object
dtype: object

Missing Values:
No missing values.

Duplicate rows: 0
Unique O*NET-SOC Codes: 1016 / 1016


---
## 4. Essential Skills Dataset (`essential_skills.csv`)
Requirements table mapping O*NET-SOC codes to essential role skills and importance ratings.


In [5]:
ess_df = pd.read_csv(os.path.join(RAW_DIR, "essential_skills.csv"))
print(f"Shape: {ess_df.shape} (Rows: {ess_df.shape[0]}, Cols: {ess_df.shape[1]})")
print("\nColumns & Dtypes:")
print(ess_df.dtypes)

print("\nMissing Values:")
missing = ess_df.isnull().sum()
print(missing[missing > 0])

print(f"\nDuplicate rows: {ess_df.duplicated().sum()}")
print(f"Unique O*NET-SOC Codes: {ess_df['O*NET-SOC Code'].nunique()}")
print(f"Unique Skill Names ('Element Name'): {ess_df['Element Name'].nunique()}")


Shape: (18200, 15) (Rows: 18200, Cols: 15)

Columns & Dtypes:
O*NET-SOC Code         object
Title                  object
Element ID             object
Element Name           object
Scale ID               object
Scale Name             object
Data Value            float64
N                       int64
Standard Error        float64
Lower CI Bound        float64
Upper CI Bound        float64
Recommend Suppress     object
Not Relevant           object
Date                   object
Domain Source          object
dtype: object

Missing Values:
Not Relevant    9100
dtype: int64

Duplicate rows: 0
Unique O*NET-SOC Codes: 910
Unique Skill Names ('Element Name'): 10


---
## 5. Software Skills Dataset (`software_skills.csv`)
Specific software tools, programming languages, and technologies per O*NET role.


In [6]:
soft_df = pd.read_csv(os.path.join(RAW_DIR, "software_skills.csv"))
print(f"Shape: {soft_df.shape} (Rows: {soft_df.shape[0]}, Cols: {soft_df.shape[1]})")
print("\nColumns & Dtypes:")
print(soft_df.dtypes)

print("\nMissing Values:")
missing = soft_df.isnull().sum()
print(missing[missing > 0] if missing.sum() > 0 else "No missing values.")

print(f"\nDuplicate rows: {soft_df.duplicated().sum()}")
print(f"Unique O*NET-SOC Codes: {soft_df['O*NET-SOC Code'].nunique()}")
print(f"Unique Software Tools ('Workplace Example'): {soft_df['Workplace Example'].nunique()}")


Shape: (31821, 7) (Rows: 31821, Cols: 7)

Columns & Dtypes:
O*NET-SOC Code       object
Title                object
Workplace Example    object
Element ID           object
Element Name         object
Hot Technology       object
In Demand            object
dtype: object

Missing Values:
No missing values.

Duplicate rows: 0
Unique O*NET-SOC Codes: 923
Unique Software Tools ('Workplace Example'): 8753


---
## Data Understanding Summary Table

| Dataset | Rows | Cols | Primary Key Candidate | Missing % | Purpose | Can Join To |
|---|---|---|---|---|---|---|
| `employee_attrition.csv` | 1,470 | 35 | `EmployeeNumber` | 0.0% | Target attrition prediction dataset | `hr_performance_engagement.csv` on Employee ID |
| `hr_performance_engagement.csv` | 2,845 | 28 | `Employee ID` | 0.0% | Engagement & survey metrics | `employee_attrition.csv` on Employee ID |
| `occupation_data.csv` | 1,016 | 3 | `O*NET-SOC Code` | 0.0% | Master occupation titles & descriptions | `essential_skills.csv`, `software_skills.csv` on SOC Code |
| `essential_skills.csv` | 18,200 | 15 | Composite (`O*NET-SOC Code`, `Element ID`, `Scale ID`) | 50.0% (`Not Relevant`) | Role essential skill requirements & scores | `occupation_data.csv` on SOC Code |
| `software_skills.csv` | 31,821 | 7 | Composite (`O*NET-SOC Code`, `Workplace Example`) | 0.0% | Role software & technology tool requirements | `occupation_data.csv` on SOC Code |
